src/sentry/tasks/releasemonitor.py

```diff
@@ -0,0 +1,165 @@
+                    totals=True,
+                    limit=CHUNK_SIZE + 1,
+                    offset=offset,
+                )
+                data = result_totals["data"]
+                count = len(data)
+                more_results = count >= CHUNK_SIZE
+                offset += count
+                total_sessions = result_totals["totals"]["sessions"]
+
+            # 3. Using the sums from #2, calculate adoption rate (relevant sessions / all sessions) update the appropriate ReleaseProjectEnvironment model adopted/unadopted fields.
+            with metrics.timer(
+                "sentry.tasks.monitor_release_adoption.process_projects_with_sessions.updates"
+            ):
+                for row in data:
+                    # Runs a single query like:
+                    # SELECT xxx FROM "sentry_releaseprojectenvironment" WHERE
+                    #   (
+                    #     "sentry_releaseprojectenvironment"."project_id" = 10
+                    #     AND "sentry_releaseprojectenvironment"."release_id" = (
+                    #       SELECT
+                    #         U0."id"
+                    #       FROM
+                    #         "sentry_release" U0
+                    #       WHERE
+                    #         (
+                    #           U0."organization_id" = 4
+                    #           AND U0."version" = foo @ 1.0.0
+                    #         )
+                    #       LIMIT
+                    #         1
+                    #     )
+                    #     AND "sentry_releaseprojectenvironment"."environment_id" = (
+                    #       SELECT
+                    #         U0."id"
+                    #       FROM
+                    #         "sentry_environment" U0
+                    #       WHERE
+                    #         (
+                    #           U0."organization_id" = 4
+                    #           AND U0."project_id" = 10
+                    #           AND U0."name" = canary
+                    #         )
+                    #       LIMIT
+                    #         1
+                    #     )
+                    #   )
+                    rpe = ReleaseProjectEnvironment.objects.get(
+                        project_id=row["project_id"],
+                        release_id=Subquery(
+                            Release.objects.filter(
+                                organization=org_id, version=row["release"]
+                            ).values("id")[:1],
+                        ),
+                        environment_id=Subquery(
+                            Environment.objects.filter(
+                                organization_id=org_id,
+                                project_id=row["project_id"],
+                                name=row["environment"],
+                            ).values("id")[:1]
+                        ),
+                    )
```

Comment:
Actually wondering whether we should try and run these queries in bulk. We could potentially do something like:

```
release_env_projects = defaultdict(list)
for row in data:
    release_env_projects[(row['release'], row['environment])].append(row['project_id'])
    
for (release, environment), project_ids in release_env_projects:
    rpe = ReleaseProjectEnvironment.objects.filter(release__version=release, environment__name=environment, project_id__in=project_ids)
    ...
```
(rough idea, I guess we'd need to store some session data alongside the projects too)

I think we're most likely to have a lower number of releases/environments so I *think* this grouping will be effective. Hard to say though.

Reply:
Aah yeah that is a good point. There definitely feels like there is some room to optimize it around the idea that some of these values are repeated and selecting the same env/release multiple times is a clear inefficiency...I'll play around with your idea and see if I can get something going